src/lib/components/workflow-actions.svelte

```diff
@@ -0,0 +1,147 @@
+<script lang="ts">
+  import { tick } from 'svelte';
+
+  import { refresh } from '$lib/stores/workflow-run';
+  import { terminateWorkflow } from '$lib/services/terminate-service';
+
+  import Modal from '$holocene/modal.svelte';
+  import { coreUserStore } from '$lib/stores/core-user';
+  import Tooltip from '$lib/holocene/tooltip.svelte';
+  import SplitButton from '$lib/holocene/split-button.svelte';
+  import MenuItem from '$lib/holocene/primitives/menu/menu-item.svelte';
+  import { cancelWorkflow } from '$lib/services/workflow-service';
+  import { toaster } from '$lib/holocene/toaster.svelte';
+
+  export let workflow: WorkflowExecution;
+  export let namespace: string;
+
+  let reason = '';
+  let showTerminationConfirmation = false;
+  let showCancellationConfirmation = false;
+  let loading = false;
+
+  const showTerminationModal = () => (showTerminationConfirmation = true);
+  const hideTerminationModal = () => (showTerminationConfirmation = false);
+  const showCancellationModal = () => (showCancellationConfirmation = true);
+  const hideCancellationModal = () => (showCancellationConfirmation = false);
+
+  const handleSuccessfulTermination = async () => {
+    showTerminationConfirmation = false;
+    reason = '';
+    $refresh = Date.now();
+    await tick();
+    toaster.push({
+      message: 'Workflow terminated.',
+      variant: 'success',
+      yPosition: 'bottom',
+    });
+  };
+
+  const handleTerminationError = () => {
+    showTerminationConfirmation = false;
+    reason = '';
+    toaster.push({ message: 'Cannot terminate workflow.', variant: 'error' });
+  };
+
+  const terminate = () => {
+    if (!workflow.canBeTerminated) return;
+    terminateWorkflow({
+      workflow,
+      namespace,
+      reason,
+    })
+      .then(handleSuccessfulTermination)
+      .catch(handleTerminationError);
+  };
+
+  const cancel = async () => {
+    loading = true;
+    try {
+      await cancelWorkflow({
+        namespace,
+        workflowId: workflow.id,
+        runId: workflow.runId,
+      });
+
+      setTimeout(() => {
+        toaster.push({
+          message: 'Canceled workflow.',
+          yPosition: 'bottom',
+        });
+        showCancellationConfirmation = false;
+        loading = false;
+        $refresh = Date.now();
+      }, 1000);
```

Comment:
Just calling out this atrocity, but it works in my testing. Should we poll instead?

Reply:
So the one thing I think worth check with OSS Server team is if a cancel request could be blocked or delayed by a current activity. I had a workflow with an WorkflowTaskFailed (see recording), and when I cancel it doesn't cancel (still running). I don't get any error from the cancel request, but maybe it does cancel just not until some task finishes.